File: src/Hostingaffe.Domain/Deployments/Deployment.cs
using Hostingaffe.Domain.History;

namespace Hostingaffe.Domain.Deployments;

/// <summary>
/// The record that an installation changed version (<c>CONTEXT.md</c>,
/// Deployment; VISION 7). It is the history of the installation and the source
/// of its current version — and it is the reason the installation carries no
/// version field of its own.
/// </summary>
/// <remarks>
/// <para>
/// <strong>There is no status.</strong> A deployment is recorded when it is
/// done. A rollback is a deployment to the previous version with a note that
/// says so; an attempt that changed nothing is a note on the installation or a
/// ticket, not a row here. That keeps the list honest: every row is a version
/// that really ran.
/// </para>
/// <para>
/// <see cref="At"/> is when the version went live and may be set, so that
/// history can be backfilled. <see cref="CreatedBy"/> — <c>by</c> on the wire —
/// is who <em>recorded</em> it, which for a backfilled deployment is not
/// necessarily who deployed.
/// </para>
/// <para>
/// <strong>The correction rule is fixed</strong>, because an agent will record
/// the wrong thing: <see cref="Ref"/>, <see cref="Ticket"/>, <see cref="Note"/>
/// and <see cref="At"/> change and the history says so; <see cref="Version"/>
/// and the installation do not, because they are what the record <em>is</em>. A
/// deployment with the wrong version is deleted and recorded again.
/// </para>
/// </remarks>
public sealed class Deployment
{
    public const int VersionMaxLength = 200;

    /// <summary>What an image reference with a digest, or a git SHA, fits in.</summary>
    public const int RefMaxLength = 500;

    /// <summary>What a planaffe key fits in.</summary>
    public const int TicketMaxLength = 64;

    private Deployment()
    {
        // EF Core materializes through this; every other route goes through Record.
    }

    private Deployment(
        Guid id, Guid installationId, int number, string version, DateTimeOffset at, Guid by, DateTimeOffset recordedAt)
    {
        Id = id;
        InstallationId = installationId;
        Number = number;
        Version = version;
        At = at;
        Note = string.Empty;
        CreatedBy = by;
        CreatedAt = recordedAt;
        UpdatedBy = by;
        UpdatedAt = recordedAt;
    }

    public Guid Id { get; private init; }

    /// <summary>Whose version changed. It does not move: that is what the record is.</summary>
    public Guid InstallationId { get; private init; }

    /// <summary>
    /// What the instance numbered it, counted from one per installation. A
    /// deployment has no key; this is what the API and the CLI address it by.
    /// </summary>
    public int Number { get; private init; }

    /// <summary>What ran afterwards — <c>2.11.4</c>, <c>main-20260905</c>. It does not change.</summary>
    public string Version { get; private init; } = null!;

    /// <summary>The exact thing that was deployed: an image reference with a digest, a git SHA, a tag.</summary>
    public string? Ref { get; private set; }

    /// <summary>When the version went live. Settable, so that history can be backfilled.</summary>
    public DateTimeOffset At { get; private set; }

    /// <summary>A planaffe key like <c>LOG-42</c>: a reference to the other product, not a word of this model.</summary>
    public string? Ticket { get; private set; }

    /// <summary>Markdown: why, what was checked, what went wrong.</summary>
    public string Note { get; private set; } = null!;

    /// <summary>Who recorded it — <c>by</c> on the wire.</summary>
    public Guid CreatedBy { get; private init; }

    /// <summary>When it was recorded, which for a backfilled deployment is not <see cref="At"/>.</summary>
    public DateTimeOffset CreatedAt { get; private init; }

    public Guid UpdatedBy { get; private set; }

    /// <summary>The version a guarded write is compared against.</summary>
    public DateTimeOffset UpdatedAt { get; private set; }

    public DateTimeOffset? DeletedAt { get; private set; }

    public Guid? DeletedBy { get; private set; }

    public bool Deleted => DeletedAt is not null;

    /// <exception cref="ArgumentException">The version does not hold.</exception>
    public static Deployment Record(
        Guid installationId, int number, string? version, DateTimeOffset at, Guid by, DateTimeOffset recordedAt) =>
        new(Guid.CreateVersion7(), installationId, number, NormalizeVersion(version), at, by, recordedAt);

    /// <summary>
    /// The four fields that may be corrected, and what actually changed.
    /// </summary>
    /// <exception cref="ArgumentException">A value does not hold; the message names the field.</exception>
    public IReadOnlyList<FieldChange> Apply(DeploymentEdit edit, Guid by, DateTimeOffset now)
    {
        ArgumentNullException.ThrowIfNull(edit);

        var changes = new List<FieldChange>();

        Fields.Text("ref", edit.Ref, Ref, value => Ref = value, value => Fields.Line(value, RefMaxLength, "A ref"), changes);
        Fields.Text("ticket", edit.Ticket, Ticket, value => Ticket = value, value => Fields.Line(value, TicketMaxLength, "A ticket"), changes);

        if (edit.At is { } moment && moment != At)
        {
            changes.Add(new FieldChange("at", Fields.Stamp(At), Fields.Stamp(moment)));
            At = moment;
        }

        // A note records that it changed, never how: it is Markdown, and the
        // text itself is one read away.
        if (edit.Note is not null && edit.Note != Note)
        {
            changes.Add(new FieldChange("note", null, null));
            Note = edit.Note;
        }

        if (changes.Count > 0)
        {
            UpdatedBy = by;
            UpdatedAt = now;
        }

        return changes;
    }

    /// <summary>Soft, with the grace period of everything else (ADR 0013).</summary>
    public void Delete(Guid by, DateTimeOffset at)
    {
        if (Deleted)
        {
            return;
        }

        DeletedAt = at;
        DeletedBy = by;
    }

    public void Restore()
    {
        DeletedAt = null;
        DeletedBy = null;
    }

    /// <exception cref="ArgumentException"><paramref name="version"/> is blank, spans lines, or is too long.</exception>
    public static string NormalizeVersion(string? version)
    {
        var trimmed = version?.Trim();

        return string.IsNullOrEmpty(trimmed)
            ? throw new ArgumentException("A deployment is a version that ran; it says which.", nameof(version))
            : Fields.Line(trimmed, VersionMaxLength, "A version");
    }
}

/// <summary>
/// What a caller wants a deployment to say afterwards. Only the four fields a
/// correction may touch are here — <c>version</c> and the installation are what
/// the record is, and a deployment with the wrong version is deleted and
/// recorded again (VISION 7).
/// </summary>
public sealed record DeploymentEdit
{
    public string? Ref { get; init; }

    public DateTimeOffset? At { get; init; }

    public string? Ticket { get; init; }

    public string? Note { get; init; }
}

File: src/cli/internal/cmd/deployment.go
package cmd

import (
	"bytes"
	"strconv"

	"github.com/spf13/cobra"

	"github.com/datavisionzero/hostingaffe/src/cli/internal/api"
	"github.com/datavisionzero/hostingaffe/src/cli/internal/client"
	"github.com/datavisionzero/hostingaffe/src/cli/internal/config"
	"github.com/datavisionzero/hostingaffe/src/cli/internal/render"
)

// The object is the deployment (CONTEXT.md); `deploy` is the short form
// VISION 6.1 types, and recording one is the bare verb, because that is what an
// agent does after the work: `ha deploy logaffe-prod --version 1.4.0`.
func newDeployment(g *globals) *cobra.Command {
	var write deploymentFields
	cmd := &cobra.Command{
		Use:     "deployment KEY --version VERSION",
		Short:   "Deployments: the record that an installation changed version. Recording one is the bare verb.",
		Aliases: []string{"deploy", "deployments"},
		Args:    cobra.MaximumNArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			if len(args) == 0 {
				return cmd.Help()
			}
			return recordDeployment(g, cmd, args[0], &write)
		},
	}
	write.flags(cmd)
	cmd.Flags().StringVar(&write.version, "version", "", "the version that is now running; the one field a deployment must have")
	cmd.AddCommand(
		newDeploymentList(g), newDeploymentView(g), newDeploymentSet(g),
		newDeploymentDelete(g), newDeploymentRestore(g), newDeploymentHistory(g))
	return cmd
}

func printDeployment(g *globals, cmd *cobra.Command, deployment api.Deployment) error {
	if g.json {
		return render.JSON(cmd.OutOrStdout(), deployment)
	}
	render.Deployment(cmd.OutOrStdout(), deployment)
	return nil
}

// deploymentFields are what a deployment is written with. `version` is on the
// recording verb only, and `set` carries its own so that the instance can say
// why it is refused rather than ha turning it away as an unknown flag.
type deploymentFields struct {
	version, ref, at, ticket string
	note, noteFile           string
}

func (d *deploymentFields) flags(cmd *cobra.Command) {
	f := cmd.Flags()
	f.StringVar(&d.ref, "ref", "", "what was actually deployed: an image digest, a commit, a tag")
	f.StringVar(&d.at, "at", "", "when it happened, if not now: 2026-09-05, or an RFC 3339 timestamp")
	f.StringVar(&d.ticket, "ticket", "", "the planaffe ticket it was done for")
	f.StringVar(&d.note, "note", "", "why, what was checked, what went wrong — the deployment's own field, in Markdown")
	f.StringVar(&d.noteFile, "note-file", "", "the note, from a file or `-` for stdin")
}

func (d *deploymentFields) body(cmd *cobra.Command) (*fields, error) {
	f := given(cmd)
	for flag, value := range map[string]*string{
		"version": &d.version, "ref": &d.ref, "ticket": &d.ticket, "note": &d.note,
	} {
		f.take(flag, value)
	}

	if err := f.takeMoment("at", &d.at); err != nil {
		return nil, err
	}

	if cmd.Flags().Changed("note") && cmd.Flags().Changed("note-file") {
		return nil, &config.UsageError{Message: "the note comes from one place: --note or --note-file, not both."}
	}

	text, err := readText(cmd.InOrStdin(), d.noteFile)
	if err != nil {
		return nil, err
	}
	if text != nil {
		f.set("note", *text)
	}

	return f, nil
}

func recordDeployment(g *globals, cmd *cobra.Command, key string, write *deploymentFields) error {
	body, err := write.body(cmd)
	if err != nil {
		return err
	}

	_, c, err := g.load()
	if err != nil {
		return err
	}
	resp, err := c.RecordDeploymentWithBodyWithResponse(
		cmd.Context(), key, "application/json", bytes.NewReader(body.json()))
	if err != nil {
		return client.Transport(err)
	}
	if err := client.Check(resp.HTTPResponse, resp.Body); err != nil {
		return err
	}
	return printDeployment(g, cmd, *resp.JSON201)
}

// newDeploymentList names the installation with a flag rather than a position,
// because the position is the recording verb's: `ha deploy KEY` is what an
// agent types after the work, and `list` does not compete with it.
func newDeploymentList(g *globals) *cobra.Command {
	var installation string
	cmd := &cobra.Command{
		Use: "list --inst KEY", Short: "Every deployment of the installation, newest by `at` first.", Args: cobra.NoArgs,
		RunE: func(cmd *cobra.Command, _ []string) error {
			if installation == "" {
				return &config.UsageError{Message: "a deployment lives under an installation: --inst KEY."}
			}
			_, c, err := g.load()
			if err != nil {
				return err
			}
			resp, err := c.ListDeploymentsWithResponse(cmd.Context(), installation)
			if err != nil {
				return client.Transport(err)
			}
			if err := client.Check(resp.HTTPResponse, resp.Body); err != nil {
				return err
			}
			if g.json {
				return render.JSON(cmd.OutOrStdout(), resp.JSON200)
			}
			render.DeploymentSummaries(cmd.OutOrStdout(), *resp.JSON200)
			return nil
		},
	}
	cmd.Flags().StringVar(&installation, "inst", "", "the installation, by `key`")
	return cmd
}

func newDeploymentView(g *globals) *cobra.Command {
	return &cobra.Command{
		Use: "view KEY NUMBER", Short: "The deployment, with the version before it and the file revisions that were current when it went live.", Args: cobra.ExactArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			number, err := aNumber(args[1])
			if err != nil {
				return err
			}
			_, c, err := g.load()
			if err != nil {
				return err
			}
			resp, err := c.ReadDeploymentWithResponse(cmd.Context(), args[0], number)
			if err != nil {
				return client.Transport(err)
			}
			if err := client.Check(resp.HTTPResponse, resp.Body); err != nil {
				return err
			}
			return printDeployment(g, cmd, *resp.JSON200)
		},
	}
}

// newDeploymentSet is narrow, and the border is the instance's: `ref`, `at`,
// `ticket` and `note`. `--version` is here so that the attempt is answered by
// the instance, which says the rule — a deployment with the wrong version is
// deleted and recorded again — rather than by ha turning the flag away.
func newDeploymentSet(g *globals) *cobra.Command {
	var write deploymentFields
	var ifMatch string
	cmd := &cobra.Command{
		Use: "set KEY NUMBER", Short: "Correct ref, at, ticket or note. The version and the installation are what the record is.", Args: cobra.ExactArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			number, err := aNumber(args[1])
			if err != nil {
				return err
			}
			body, err := write.body(cmd)
			if err != nil {
				return err
			}
			if body.nothing() {
				return &config.UsageError{Message: "nothing to correct: --ref, --at, --ticket, --note or --note-file."}
			}

			_, c, err := g.load()
			if err != nil {
				return err
			}
			resp, err := c.CorrectDeploymentWithBodyWithResponse(
				cmd.Context(), args[0], number, "application/json", bytes.NewReader(body.json()), guard(ifMatch))
			if err != nil {
				return client.Transport(err)
			}
			if err := client.Check(resp.HTTPResponse, resp.Body); err != nil {
				return err
			}
			return printDeployment(g, cmd, *resp.JSON200)
		},
	}
	write.flags(cmd)
	cmd.Flags().StringVar(&write.version, "version", "", "refused: the version is what the record is — delete it and record it again")
	ifMatchFlag(cmd, &ifMatch)
	return cmd
}

// newDeploymentDelete is the other half of the correction rule: a deployment
// recorded with the wrong version is deleted and recorded again. Its number is
// not handed out a second time.
func newDeploymentDelete(g *globals) *cobra.Command {
	return &cobra.Command{
		Use: "delete KEY NUMBER", Short: "Soft-delete a deployment. Its number is not handed out again.", Args: cobra.ExactArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			number, err := aNumber(args[1])
			if err != nil {
				return err
			}
			_, c, err := g.load()
			if err != nil {
				return err
			}
			resp, err := c.DeleteDeploymentWithResponse(cmd.Context(), args[0], number)
			if err != nil {
				return client.Transport(err)
			}
			if err := client.Check(resp.HTTPResponse, resp.Body); err != nil {
				return err
			}
			if g.json {
				return render.JSON(cmd.OutOrStdout(), map[string]any{
					"installation": args[0], "number": number, "deleted": true})
			}
			_, _ = cmd.OutOrStdout().Write([]byte(
				args[0] + " #" + args[1] + " deleted; `ha deploy restore " + args[0] + " " + args[1] + "` brings it back.\n"))
			return nil
		},
	}
}

func newDeploymentRestore(g *globals) *cobra.Command {
	return &cobra.Command{
		Use: "restore KEY NUMBER", Short: "Bring a deleted deployment back, under the number it kept.", Args: cobra.ExactArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			number, err := aNumber(args[1])
			if err != nil {
				return err
			}
			_, c, err := g.load()
			if err != nil {
				return err
			}
			resp, err := c.RestoreDeploymentWithResponse(cmd.Context(), args[0], number)
			if err != nil {
				return client.Transport(err)
			}
			if err := client.Check(resp.HTTPResponse, resp.Body); err != nil {
				return err
			}
			return printDeployment(g, cmd, *resp.JSON200)
		},
	}
}

// newDeploymentHistory is the corrections made to one deployment. The
// deployments themselves are not history entries — they are records of their
// own (CONTEXT.md, Deployment).
func newDeploymentHistory(g *globals) *cobra.Command {
	return &cobra.Command{
		Use: "history KEY NUMBER", Short: "Every correction made to the deployment, oldest first.", Args: cobra.ExactArgs(2),
		RunE: func(cmd *cobra.Command, args []string) error {
			number, err := aNumber(args[1])
			if err != nil {
				return err
			}
			_, c, err := g.load()
			if err != nil {
				return err
			}
			resp, err := c.ReadDeploymentHistoryWithResponse(cmd.Context(), args[0], number)
			if err != nil {
				return client.Transport(err)
			}
			if err := client.Check(resp.HTTPResponse, resp.Body); err != nil {
				return err
			}
			if g.json {
				return render.JSON(cmd.OutOrStdout(), resp.JSON200)
			}
			render.History(cmd.OutOrStdout(), *resp.JSON200)
			return nil
		},
	}
}

// aNumber is the address of a deployment: the instance numbers it per
// installation, and anything that is not one of those numbers is a usage
// mistake said before any request goes out.
func aNumber(text string) (int32, error) {
	number, err := strconv.ParseInt(text, 10, 32)
	if err != nil || number < 1 {
		return 0, &config.UsageError{
			Message: "a deployment is addressed by its number, counting from 1: `ha deploy list --inst KEY` says which.",
		}
	}
	return int32(number), nil
}
